Enabled several setoption UCI commands:

TheAnswer opened the pull request:
Added the following setoption name <name> value <value> UCI options:

setoption name threads value <int>
setoption name playouts value <int>
setoption name visits value <int>
setoption name quiet value <int> 0 or 1
setoption name fpu_dynamic_eval value <float>
setoption name fpu_reduction value <float>
setoption name softmax_temp value <float>
setoption name puct value <float>

evalon32 commented:
AFAICT, the engine needs to advertise the options at startup, so the GUI will know what to display. I imagine you'll want to send something like this:
```
option name threads type spin default $cfg_num_threads min 1 max $cfg_max_threads
option name playouts type spin default $cfg_max_playouts min 0 max $INT_MAX
option name visits type spin default $cfg_max_visits min 0 max $INT_MAX
option name quiet type check default false
option name fpu_dynamic_eval type string default $cfg_fpu_dynamic_eval
etc.
```
Also, it looks like the convention is to capitalize option names. They are typically used verbatim for display in GUI.

killerducky commented:
What about a division of options into those we advertise through UCI to the GUI and those we don't. For example fpu_dynamic_eval could be an unadvertised option. But maybe that would make it more difficult for those doing the tuning to use? What do people think?

TheAnswer commented:
Ok I missed that part. 

I will rework it then into a proper map which will filter the commands to advertise as needed. 
That will make it easier to add more commands later on instead of these hardcoded if-checks and functions.

Tilps commented:
I'm not sure playouts and visits should be options.  go nodes x should be the way to perform fixed visits.
(Maybe there could be an option to decide whether x in go nodes x is playouts or visits...)

Ipmanchess commented:
With UCI engines we get this as a example:

DON 160418.64.ABM.LP by Ehsan Rashid
info string Processor(s) detected 4
info string Normal Pages Hash 16 MB
info string Thread(s) used 1
uci
id name DON 160418.64.ABM.LP
id author Ehsan Rashid
option name Hash type spin default 16 min 0 max 131072
option name Large Pages type check default true
option name Clear Hash type button
option name Retain Hash type check default false
option name Hash File type string default Hash.dat
option name Save Hash type button
option name Load Hash type button
option name Use Book type check default false
option name Book File type string default Book.bin
option name Book Pick Best type check default true
option name Book Move Count type spin default 20 min 0 max 100
option name Threads type spin default 1 min 0 max 512
option name Skill Level type spin default 12 min 0 max 12
option name MultiPV type spin default 1 min 1 max 255
option name Fixed Contempt type spin default 0 min -100 max 100
option name Timed Contempt type spin default 60 min 0 max 1000
option name Valued Contempt type spin default 10 min 0 max 1000
option name Draw MoveCount type spin default 50 min 5 max 50
option name Overhead Move Time type spin default 30 min 0 max 5000
option name Minimum Move Time type spin default 20 min 0 max 5000
option name Move Slowness type spin default 84 min 10 max 1000
option name Nodes Time type spin default 0 min 0 max 10000
option name Ponder type check default true
option name SyzygyPath type string default <empty>
option name SyzygyProbeDepth type spin default 1 min 1 max 100
option name SyzygyLimitPiece type spin default 6 min 0 max 6
option name SyzygyUseRule50 type check default true
option name Debug File type string default
option name Output File type string default <empty>
option name UCI_Chess960 type check default false
option name UCI_LimitStrength type check default false
option name UCI_ELO type spin default 3490 min 1250 max 3490
uciok

Komodo 2041.00 64-bit (C) 2018 Don Dailey, Larry Kaufman and Mark Lefler
using hardware POPCNT
info string Licensed to Ipmanchess
uci
id name Komodo 2041.00 64-bit
id author Don Dailey, Larry Kaufman, Mark Lefler
option name Threads type spin default 1 min 1 max 64
option name Ponder type check default true
option name Hash type spin default 192 min 1 max 65536
option name Clear Hash type button
option name Table Memory type spin default 128 min 1 max 1024
option name MultiPV type spin default 1 min 1 max 218
option name OwnBook type check default true
option name Book Moves type spin default 1000 min 0 max 1000
option name Book File type string default komodo.bin
option name Best Book Line type check default false
option name Log File type string default <empty>
option name Hash File Name type string default <empty>
option name Save Hash to File type button
option name Load Hash from File type button
option name UCI_Chess960 type check default false
option name UCI_AnalyseMode type check default false
option name Use LMR type check default true
option name Null Move Pruning type check default true
option name Minimal Reporting type spin default 0 min 0 max 200
option name Overhead ms type spin default 30 min 0 max 10000
option name Time Usage type spin default 0 min -9 max 9
option name Use Syzygy type check default true
option name SyzygyPath type string default <empty>
option name Syzygy Probe Depth type spin default 2 min 0 max 99
option name Syzygy Probe Limit type spin default 6 min 0 max 6
option name Syzygy 50 Move Rule type check default true
option name Smart Syzygy type check default false
option name King Safety type spin default 70 min 0 max 200
option name Selectivity type spin default 130 min 10 max 250
option name Reduction type spin default -20 min -1000 max 120
option name Dynamism type spin default 130 min 0 max 400
option name Contempt type spin default 16 min -500 max 500
option name White Contempt type check default false
option name Progress Threshold type spin default 30 min 10 max 99
option name Variety type spin default 0 min 0 max 20
option name Skill type spin default 20 min 1 max 20
uciok

Stockfish 9 64 POPCNT by T. Romstad, M. Costalba, J. Kiiski, G. Linscott
uci
id name Stockfish 9 64 POPCNT
id author T. Romstad, M. Costalba, J. Kiiski, G. Linscott

option name Debug Log File type string default
option name Contempt type spin default 20 min -100 max 100
option name Threads type spin default 1 min 1 max 512
option name Hash type spin default 16 min 1 max 131072
option name Clear Hash type button
option name Ponder type check default false
option name MultiPV type spin default 1 min 1 max 500
option name Skill Level type spin default 20 min 0 max 20
option name Move Overhead type spin default 30 min 0 max 5000
option name Minimum Thinking Time type spin default 20 min 0 max 5000
option name Slow Mover type spin default 89 min 10 max 1000
option name nodestime type spin default 0 min 0 max 10000
option name UCI_Chess960 type check default false
option name SyzygyPath type string default <empty>
option name SyzygyProbeDepth type spin default 1 min 1 max 100
option name Syzygy50MoveRule type check default true
option name SyzygyProbeLimit type spin default 6 min 0 max 6
uciok

TheAnswer commented:
I moved the uci options implementation from stockfish seeing that parts of it was already in the code.

I also added SilentOption to disable advertising for some of them. 

The option list so far then is:

setoption name Threads value <int>
setoption name Quiet value <int> 0 or 1
setoption name Fpu Dynamic Eval value <float>
setoption name Fpu Reduction value <float>
setoption name Softmax Temp value <float>
setoption name Puct value <float>

Fpu Dynamic Eval and Softmax Temp are silent options and not advertised in the uci command.

evalon32 commented on src/UCIOption.cpp:
```diff
@@ -0,0 +1,202 @@
+    void on_softmaxtemp(const Option& o) {
+        std::string value = o;
+
+        auto val = static_cast<float>(std::atof(value.c_str()));
```
std::stof? (same below)

TheAnswer replied:
Switched

jjoshua2 commented:
Could you make slow mover a uci option please? In the timemanage.cpp I think?

TheAnswer commented:
Switched to strtof and added SlowMover option

TheAnswer commented:
Also added Go Nodes Visits where true sets go nodes to vists and is the default, false sets go nodes to playouts

killerducky commented:
The threads option does not appear, not sure why? I'm using Arena on linux.

evalon32 commented:
@killerducky I think "threads" is treated specially -- it doesn't appear for Stockfish, either. Try "Common max CPU cores setting" under common UCI options.

killerducky commented:
@evalon32 Ok I found it -- Engines, Manage, UCI tab. It's working! It was default 1 (ugh??), bumped it to 4 and ran, down to 1 ran, up to 4 ran, all working as expected.

@TheAnswer any last minute changes before I merge this?
